crates/binjs_io/src/entropy/model.rs

```diff
@@ -1,183 +1,251 @@
+        + self.string_literal_by_path.len()
+        + self.list_length_by_path.len()
     }
 }
 
+/// Counting instances of each kind of string in a file.
+///
+/// We use this data to determine which strings should go in a shared dictionary
+/// or which should be inlined in the file.
+#[derive(Debug, Default)]
+pub struct StringsInFile<T> {
+    /// Instances of IdentifierName.
+    pub identifier_name_instances: HashMap<Option<IdentifierName>, T>,
+
+    /// Instances of PropertyKey
+    pub property_key_instances: HashMap<Option<PropertyKey>, T>,
+
+    /// Instances of InterfaceName
+    pub interface_name_instances: HashMap<InterfaceName, T>,
+
+    /// Instances of string literals.
+    pub string_literal_instances: HashMap<Option<SharedString>, T>,
 
-/// An encoding model which starts by analyzing the full AST to determine
-/// exact statistics.
-pub struct ExactEncodingModel {
-    probabilities: ExactEncodingModelData<Symbol>,
+    /// Instances of string enums.
+    pub string_enum_instances: HashMap<SharedString, T>,
 }
-impl ExactEncodingModel {
-    fn get_from_path<T>(predictor: &mut PathPredict<T, Symbol>, value: &T, path: &ASTPath) -> Result<Symbol, ()>
-        where T: Eq + std::hash::Hash + Clone + std::fmt::Debug
-    {
-        let segment = predictor.get_mut(path, value)
-            .ok_or(())?;
-        let result = segment.clone();
-        Ok(result)
-    }
+
+/// A structure used to build a dictionary based on a sample of files.
+///
+/// In this version, we use several dictionary lengths, to determine the
+/// best version for a file.
+pub struct DictionaryBuilder<'a> {
+    /// Attempt to build predictions with a depth of 0 (no context) up to max_path_depth ancestors.
+    max_path_depth: usize,
+
+    /// A dictionary.
+    dictionary: &'a mut Dictionary</* instances */ usize>,
+
+    /// Number of instances of each string in the current file.
+    instances_of_strings_in_current_file: StringsInFile<usize>,
+
+    /// Number of files in which each string appears.
+    files_containing_string: &'a mut StringsInFile<usize>,
 }
-impl EncodingModel for ExactEncodingModel {
-    fn string_frequency_for_encoding(&mut self, value: &Option<SharedString>, path: &ASTPath) -> Result<Symbol, ()> {
-        Self::get_from_path(&mut self.probabilities.strings, value, path)
+
+impl<'a> DictionaryBuilder<'a> {
+    pub fn new(max_path_depth: usize, dictionary: &'a mut Dictionary<usize>, files_containing_string: &'a mut StringsInFile<usize>) -> Self {
+        DictionaryBuilder {
+            max_path_depth,
+            dictionary,
+            instances_of_strings_in_current_file: StringsInFile::default(),
+            files_containing_string
+        }
+    }
+
+    fn add_instance_to_path<V>(max_path_depth: usize, value: V, path: &IOPath, source: &mut PathPredict<V, usize>)
+        where
+            V: 'a + std::hash::Hash + Eq + Clone + std::fmt::Debug,
+    {
+        let len = usize::min(max_path_depth, path.len()) + 1;
```

Comment:
with max_path_depth is parameter here, instead of some constant?

Reply:
Because I'm planning to make it a command-line argument.

Reply:
looks like, `add_instance_to_path` and `add_instance_to_strings` are defined like static method, but actually receives `DictionaryBuilder` fields.
can we just make them instance method and receive `self` and extract fields from it ?
that way it's clearer what this method does.